private/buf/bufgen/internal/bufgenv2/file_option.go

```diff
@@ -59,8 +58,21 @@ const (
 	fileOptionPhpMetadataNamespaceSuffix
 	// fileOptionRubyPackage is the file option ruby_package.
 	fileOptionRubyPackage
+	// fileOptionRubyPackageSuffix is the file option ruby_package_suffix.
+	fileOptionRubyPackageSuffix
 	// groupJavaPackage is the file option group that modifies java_package.
 	groupJavaPackage fileOptionGroup = iota + 1
+	groupJavaOuterClassname
```

Comment:
Can you quickly explain why we need groups for all of these?

Reply:
There can be multiple `file_option`s in `buf.gen.yaml` that modify the same file option on an image file's descriptor set (e.g. `java_package`, `java_package_prefix` and `java_package_suffix` all modify java package). The order among rules with different `file_options` matters if they modify the same file option.
For example:
```
override: ## values omitted
  - file_option: java_package
  - file_option: java_package_prefix
 ```
and
```
override:
  - file_option: java_package_prefix
  - file_option: java_package
```
have different effects.
This is achieved by putting override in different groups and for each group, before we call `mergeOverrideFunc`. [`mergeOverrideFuncs` can then decide what happens if the last two are prefix and suffix, or prefix and value and etc.](https://github.com/bufbuild/buf/pull/2344/files#diff-6ad607dc37b36d1f3e43504fb3b079819f8d2cca6cb92f7b16c2714ef2b64e86R404)

Another approach is to attach each rule with a number (the position it appears in the list of override rules).